docs/src/examples/matrix-inversion-manual.jl

```diff
@@ -143,9 +143,11 @@ MOI.set(
     model,
     DiffOpt.ForwardInConstraint(),
     cons[1],
-    0.0 * index(x[1]) - 1.0,
+    0.0 * index(x[1]) - 1.0,  # the tangent of the ConstraintFunction, i.e., ∂(Gx - h)/∂h = -1
```

Comment:
Not exactly.
This is actually a perturbation on the problem. That is the direction that the derivative will be computed. In this case the direction is only following the "h[1]" axis and nothing else. There is one axis for each parameter of the problem "h[1]", "G[1,1]", "G[1,2]" and so on.

Then, the get ForwardOutVariablePrimal will return the derivative of 'x' in that direction

Reply:
Thank you for your comment.
I'm really sorry that I don't get it clearly.

Can you explain why the argument should be `0.0 * index(x[1]) - 1.0` and how it changes if different parameter is taken for calculation of derivatives?

Reply:
Apart from that, I think that the output of `get ForwardOutVariablePrimal` returns `dx/dh` in the tutorial, right?

Reply:
Derivatives in Rn require a direction, see:
https://en.wikipedia.org/wiki/Directional_derivative

that argument is defining that direction to be (1 * h[1], 0*G[1,1], 0*G[1,2], ...).

In other (very different) words, that carries the information of which vector is going to be multiplied by the Jacobian: (1 * dh[1], 0*dG[1,1], 0*dG[1,2], ...)

Reply:
see also: https://math.stackexchange.com/questions/3191003/directional-derivative-and-jacobian-matrix

Reply:
Thank you, @joaquimg.
But my question is, in that line, "how to let the code know the direction" for directional derivative calculation.
That is,
I don't understand how the line `0.0 * index(x[1]) - 1.0` indicates that "the direction is (1, 0, 0, ...) with axes of (h[1], G[1, 1], G[1, 2], ...)".

Reply:
What if I want to decide the direction as (0, 1, 0) with axes of (h[1], G[1, 1], G[1, 2])?

Reply:
In that case you would do:
1.0x[1] + 0.0

cons[1] means that you are selecting the first constraint and x[1] means the coefficient of the first variable: that is G[1,1].

The 1.0 multiplying x[1] means 1 in that direcation.

Reply:
Well, I see.
Then, you mean that when considering the constraint `Gx - h <=0`, the direction of `G[1, i]` corresponds to `x[i]` and `h` corresponds to the constant...?
Also,
1) the `-1.0` in `0.0 * index(x[1]) - 1.0` corresponds to "1 in h-direction", or "-h"-direction? (it should be "h-direction" in context, probably due to the negative sign convention of `h` in `Gx - h <= 0`.
2) `0.0 * index(x[1])` is still to make the type a AbstractScalarSomething? (I forgot the name, sorry)

Reply:
EDIT:
I'm not sure but from my observation that the `0.0 * index(x[1]) - 1.0` corresponds to `dG * x - dh`; probably that's why @joaquimg referred to as "directional derivative".
This can be found in some source codes like [how to get `dh`](https://github.com/jump-dev/DiffOpt.jl/blob/cf808b04c122a73500cdd4db9bc269b63ee3b709/src/diff_opt.jl#L689).

Please inform me if I misunderstood.

Reply:
the sign on the constant is a bit non-trivial, see my own reflections here: https://github.com/jump-dev/DiffOpt.jl/issues/174
might help...

Reply:
@joaquimg #174 provides a very good insight of how it works.
It would be better to read it before reading the source codes... 👍

Reply:
@joaquimg 
Can you check whether I understood it correctly?

The constraint function is basically the form of `f(x, G, h_) = Gx + h_` internally,
so to get the directional derivative `h`-axis (with unit vector), we have to flip the sign of the constant as we consider the constraint of `f(x, G, -h) = Gx - h`.
Also, to get directional derivatives of other axes, in the case of `G[1, i]`, just passing `x[i]` into where `0.0 * index(x[1]) - 1.0` is placed provides the corresponding directional derivative (with unit vector).
Multiplication by scalar replaces the unit vector condition intuitively.

Reply:
looks good!

Reply:
For the change request, see the following commits.